compiler/vm/irgen.nim

```diff
@@ -0,0 +1,2256 @@
+  report.location = some(loc)
+  report.reportInst = toReportLineInfo(inst)
+  raise (ref VmGenError)(report: report)
+
+func fail(
+  info: TLineInfo,
+  kind: ReportKind,
+  ast:  PNode = nil,
+  sym:  PSym = nil,
+  str:  string = "",
+  loc:  InstantiationInfo = instLoc()
+  ) {.noinline, noreturn.} =
+  raiseVmGenError(
+    SemReport(kind: kind, ast: ast, sym: sym, str: str),
+    info,
+    loc)
+
+func irSym(c: var TCtx, sym: PSym): IRIndex =
+  let id = c.defSyms.requestSym(sym)
+  c.irs.irSym(id)
+
+func irParam(c: var TCtx, sym: PSym): IRIndex =
+  c.irs.irParam(sym.position.uint32)
+
+func irGlobal(c: var TCtx, sym: PSym): IRIndex =
+  c.irSym(sym)
+
+func irConst(c: var TCtx, sym: PSym): IRIndex =
+  assert sym.kind == skConst
+  let id = c.defSyms.requestSym(sym)
+
+  if not c.seenConsts.containsOrIncl(id):
+    # XXX: collecting constants should *not* be the responsibility of
+    #      ``irgen``. But with constants still sharing their ID namespace with
+    #      globals, it's the easiest solution for now
+    c.collectedConsts.add sym
+    c.constData[id] = astdef(sym)
+
+  c.irs.irSym(id)
+
+func irLit(c: var TCtx, n: PNode): IRIndex =
+  let typ =
+    if n.typ != nil:
+      c.types.requestType(n.typ)
+    elif n.kind in nkStrLit..nkTripleStrLit:
+      # XXX: without type information, we cannot know if the string literal
+      #      is supposed to be a Nim-string or cstring, and so fall back to
+      #      unconditionally treating it as a Nim-string. Not doing so
+      #      would result in the code-generator treating it as a cstring
+      # TODO: instead of this workaround, each occurence in the compiler where
+      #       a string literal without type information is inserted needs to
+      #       corrected instead. One such occurence is ``genEnumToStrProc``.
+      {.noSideEffect.}:
+        c.types.requestType(c.graph.getSysType(unknownLineInfo, tyString))
+    else:
+      NoneType
+
+  c.irs.irLit (c.data.add(n), typ)
+
+func irImm(c: var TCtx, val: SomeInteger): IRIndex =
+  # TODO: ``ntkImm(ediate)`` needs to be used here
+  c.irs.irLit (c.data.newLit(val), c.passEnv.sysTypes[tyInt])
+
+template tryOrReturn(code): untyped =
+  try:
+    code
+  except VmGenError as e:
+    return IrGenResult.err(move e.report)
+
+func requestType(c: var TCtx, k: TTypeKind): TypeId =
+  {.cast(noSideEffect).}:
+    # XXX: ``getSysType`` has error reporting related side-effects
+    c.types.requestType(c.graph.getSysType(unknownLineInfo, k))
+
+proc getTemp(cc: var TCtx; tt: PType): IRIndex
+
+func pushScope(p: var PProc; finalizer, handler = JoinPoint(-1)) =
+  p.scopes.add Scope(firstLocal: p.activeLocals.len,
+                     finalizer: finalizer, handler: handler)
+
+func popScope(p: var PProc) =
+  # TODO: take ``PProc`` instead of ``TCtx``
+  let scope = p.scopes.pop()
+  p.activeLocals.setLen(scope.firstLocal)
+
+# XXX: the emission of ``ntkLocEnd`` instructions is disabled for now. When
```

Comment:
Since scopes (in the context of lifetimes) aren't directly encoded in the IR, the original idea was to use `ntkLocEnd` as a way to signal the end-of-life of a location on a control-flow path. This information was meant to be used by the move-analyser and related data-flow analysis, as well as the code-generator for the VM (to simplify register allocation).

Because each local can have more than one associated `ntkLocEnd` instruction, the sequence idea wouldn't work. Attaching extra out-of-band information to instructions is also a bit problematic right now, as the attachment position have to be adjusted separately on each code modification.

Aside: Instructions might need a stable ID for other things, but because of the additional memory they'd take up, I'm still trying to get around having to introduce them.

With the change of plan (i.e. the `injectdestructors` rewrite that I'll be working on first), most of the ideas around `ntkLocEnd` have become obsolete/stale.

Reply:
Thank you for through explanation.

After coming back to this after some time, I think I get it more than I did before. `ntkLocEnd` is effectively acting as a "consume" instruction being automatically inserted by the compiler. This needs to be branch aware at present... but also leads me back to thinking it simplifies under CPS. 😅 

But don't worry about all that jazz, it's just me thinking out loud.